lib/engine/game/g_18_cz/game.rb

```diff
@@ -2498,35 +2508,127 @@ def setup
+            @corporations = @corporations.reject { |item| TWO_PLAYER_CORP_TO_REMOVE.include?(item.name) }
+
+            @corporations.select { |item| item.type == :large }.each { |item| item.max_ownership_percent = 70 }
+
+            @players << @vaclav
+          end
 
           # Only small companies are available until later phases
           @corporations, @future_corporations = @corporations.partition { |corporation| corporation.type == :small }
 
+          new_corporation_for_vaclave(:small) unless multiplayer?
+
           block_lay_for_purple_tiles
           init_player_debts
         end
 
+        def new_corporation_for_vaclave(size)
+          possible_corporations = @corporations.select { |corporation| corporation.type == size }
+          index = rand % possible_corporations.size
+          new_corporation = possible_corporations[index]
+
+          new_corporation.ipoed = true
+          new_corporation.owner = @vaclav
+          new_corporation.tokens.each { |item| item.price = 0 }
+          @recently_floated << new_corporation
+          @vaclavs_corporations << new_corporation
+
+          par_value = PAR_RANGE[new_corporation.type].first
+          price = @stock_market.par_prices.find { |p| p.price == par_value }
+          @stock_market.set_par(new_corporation, price)
+
+          index = 0
+          until new_corporation.floated?
+            bundle = new_corporation.ipo_shares[index].to_bundle
+            @share_pool.transfer_shares(bundle, @vaclav)
+            index += 1
+          end
+
+          @log << "Vaclav receives new corporation #{new_corporation.name}"
+          new_train_for_vaclav(new_corporation)
+        end
+
+        def new_train_for_vaclav(corporation)
+          corporation.trains.each do |item|
+            remove_train(item)
+            item.owner = nil
+          end
+          train = @depot.upcoming.first
+          variant = train.variants.values.find { |item| train_of_size?(item, corporation.type) }
+          train.variant = variant[:name]
+          remove_train(train)
+          train.owner = corporation
+          corporation.trains << train
+
+          @phase.buying_train!(corporation, train)
+
+          @log << "#{corporation.name} receives a new #{train.name} train"
+        end
+
+        def init_companies(players)
+          companies = super
+          unless multiplayer?
+            companies = companies.reject { |item| item.value >= TWO_PLAYER_COMPANIES_TO_REMOVE[item.revenue] }
+          end
+          companies
+        end
+
+        def active_players
+          active = super
+          return active if multiplayer?
+
+          active.reject { |item| item == @vaclav }
+        end
+
+        def optional_hexes
+          return game_hexes if multiplayer?
+
+          new_hexes = {}
+          game_hexes.keys.each do |color|
+            new_map = game_hexes[color].map do |coords, tile_string|
+              [coords - TWO_PLAYER_HEXES_TO_REMOVE, tile_string]
```

Comment:
.transform_keys
cf. https://github.com/tobymao/18xx/pull/4653/commits/e5b152864eb10c368be4769220bb236e50d72104

Reply:
thanks. did not know that even exists ;)